src/engine/GroupBy.cpp

```diff
@@ -388,15 +373,125 @@ ProtoResult GroupBy::computeResult([[maybe_unused]] bool requestLaziness) {
     return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
   }
 
-  size_t inWidth = subresult->idTable().numColumns();
-  size_t outWidth = getResultWidth();
+  // TODO<RobinTF> Add runtime parameter to toggle of index scan specific
+  // optimizations for performance comparisons
+  if (subresult->isFullyMaterialized()) {
+    size_t inWidth = subresult->idTable().numColumns();
+    size_t outWidth = getResultWidth();
+
+    IdTable idTable = CALL_FIXED_SIZE(
+        (std::array{inWidth, outWidth}), &GroupBy::doGroupBy, this,
+        subresult->idTable(), groupByCols, aggregates, &localVocab);
+
+    LOG(DEBUG) << "GroupBy result computation done." << std::endl;
+    return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
+  }
 
-  IdTable idTable = CALL_FIXED_SIZE(
-      (std::array{inWidth, outWidth}), &GroupBy::doGroupBy, this,
-      subresult->idTable(), groupByCols, aggregates, &localVocab);
+  auto aggregateInfo = computeHashMapOptimizationMetadata(aggregates);
 
-  LOG(DEBUG) << "GroupBy result computation done." << std::endl;
-  return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
+  if (!aggregateInfo.has_value()) {
+    // TODO<RobinTF> Should this throw an exception?
+    return {IdTable{getResultWidth(), getExecutionContext()->getAllocator()},
+            resultSortedOn(), std::move(localVocab)};
+  }
+
+  size_t size = getResultWidth() - groupByCols.size();
+  cppcoro::generator<IdTable> generator = CALL_FIXED_SIZE(
+      size, &GroupBy::computeResultLazily, this, std::move(subresult),
+      std::move(aggregateInfo).value().aggregateAliases_,
+      std::move(groupByCols));
+  return {std::move(generator), resultSortedOn(), std::move(localVocab)};
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+cppcoro::generator<IdTable> GroupBy::computeResultLazily(
+    std::shared_ptr<const Result> subresult,
+    std::vector<HashMapAliasInformation> aggregateAliases,
+    std::vector<size_t> groupByCols) const {
+  AD_CONTRACT_CHECK(groupByCols.size() + NUM_AGGREGATED_COLS ==
+                    getResultWidth());
+
+  auto localVocab = subresult->getCopyOfLocalVocab();
+
+  LazyGroupBy<NUM_AGGREGATED_COLS> lazyGroupBy{
+      localVocab, std::move(aggregateAliases), groupByCols};
+
+  IdTable resultTable{getResultWidth(), getExecutionContext()->getAllocator()};
+
+  // TODO<RobinTF> Check if it is possible to not prolong the lifetime of this
+  // `IdTable` by seeing if the evaluation context really needs a proper
+  // `IdTable` to work.
+
+  // Extend the lifetime of the `IdTable` yielded by the generator a little bit
+  std::optional<IdTable> idTableWrapper = std::nullopt;
+  std::optional<sparqlExpression::EvaluationContext> evaluationContext =
+      std::nullopt;
```

Comment:
I don't understand yet, where the lifetime extension is needed, maybe we can talk about this.